core/state/pruner/pruner.go

```diff
@@ -332,6 +336,234 @@ func (p *Pruner) Prune(root common.Hash) error {
+
+			start = time.Now()
+		}
+	}
+
+	if err = lock.Release(); err != nil {
+		return fmt.Errorf("failed to release file lock: %v", err)
+	}
+
+	log.Info("Backup old ancientDB done", "current start blockNumber in ancientDB", startBlockNumber)
+
+	return nil
+}
+
+// BlockPruneBackup backup the ancient data for the old ancient db, i.e. the most recent 128 blocks in ancient db.
+func (p *BlockPruner) BlockPruneBackup(name string, cache, handles int, namespace string, readonly, interrupt bool) error {
+	start := time.Now()
+
+	if err := p.backupOldDb(name, cache, handles, namespace, readonly, interrupt); err != nil {
+		log.Error("Backup old ancientDB error", "err", err)
+		return err
+	}
+
+	log.Info("Block pruning backup successfully", "time duration since start is", common.PrettyDuration(time.Since(start)))
+
+	return nil
+}
+
+func (p *BlockPruner) RecoverInterruption(name string, cache, handles int, namespace string, readonly bool) error {
+	log.Info("RecoverInterruption for block prune")
+
+	newExist, err := checkFileExist(p.newAncientPath)
+	if err != nil {
+		return fmt.Errorf("newAncientDB path error %v", err)
+	}
+
+	//nolint:nestif
+	if newExist {
+		log.Info("New ancientDB_backup existed in interruption scenario")
+
+		flockOfAncientBack, err := checkFileExist(filepath.Join(p.newAncientPath, "PRUNEFLOCKBACK"))
+		if err != nil {
+			return fmt.Errorf("failed to check flock of ancientDB_Back %v", err)
+		}
+
+		// Indicating both old and new ancientDB existed concurrently.
+		// Delete directly for the new ancientDB to prune from start, e.g.: path ../chaindb/ancient_backup
+		if err := os.RemoveAll(p.newAncientPath); err != nil {
+			return fmt.Errorf("failed to remove old ancient directory %v", err)
+		}
+
+		if flockOfAncientBack {
+			// Indicating the oldOffset/newOffset have already been updated.
+			if err := p.BlockPruneBackup(name, cache, handles, namespace, readonly, true); err != nil {
+				return err
+			}
+		} else {
+			// Indicating the flock did not exist and the new offset did not be updated, so just handle this case as usual.
+			if err := p.BlockPruneBackup(name, cache, handles, namespace, readonly, false); err != nil {
+				return err
+			}
+		}
+
+		if err := p.AncientDbReplacer(); err != nil {
+			return err
+		}
+	} else {
+		log.Info("New ancientDB_backup did not exist in interruption scenario")
+		// Indicating new ancientDB even did not be created, just prune starting at backup from startBlockNumber as usual,
+		// in this case, the new offset have not been written into kvDB.
+		if err := p.BlockPruneBackup(name, cache, handles, namespace, readonly, false); err != nil {
+			return err
+		}
+		if err := p.AncientDbReplacer(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func checkFileExist(path string) (bool, error) {
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			// Indicating the file didn't exist.
+			return false, nil
```

Comment:
This is because if the file doesn't exist, it also returns an error. At the parent function, we don't really want to act upon that error. In that case, we just need to create a new file and proceed.